File: 347_Top K Frequent Elements.java
// Time complexity: O(n)
// Space complexity: O(n)

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // Bucket Sort
        // 1. count frequency
        Map<Integer, Integer> count = new HashMap<>();
        for (int i : nums) count.put(i, count.getOrDefault(i, 0) + 1);
        // 2. sort based on frequency - place into buckets
        List<Integer>[] bucket = new List[nums.length + 1];
        for (int i = 0; i <= nums.length; i++) bucket[i] = new ArrayList<>();
        for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
            int curNum = entry.getKey(), curCount = entry.getValue();
            bucket[curCount].add(curNum);
        }
        // 3. iterate backwards through buckets
        int[] res = new int[k];
        int index = 0;
        for (int i = nums.length; i > 0; i--) {
            for (int j : bucket[i]) {
                res[index] = j;
                index ++;
                if (index == k) return res;
            }
        }
        return res;
    }
}

File: 347_Top K Frequent Elements.py
# Method 1: Bucket Sort - O(n)
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # 1. count letter frequency
        count = Counter(nums)
        # 2. create len(nums) buckets and fill
        bucket = [[] for _ in range(len(nums) + 1)]
        for i, cnt in count.items():
            bucket[cnt].append(i)
        # 3. loop backwards
        res = []
        for i in range(len(nums), -1, -1):
            res += bucket[i]
            if len(res) == k: break
        return res
        
# # Method 2: Heap - O(n log(n))
# class Solution:
#     def topKFrequent(self, nums: List[int], k: int) -> List[int]:
#         # 1. count letter
#         count = Counter(nums)
#         # 2. loop through counter while maintain a heap w/ size k
#         minh = []
#         for i, cnt in count.items():
#             if len(minh) < k: heappush(minh, (cnt, i))
#             elif minh[0][0] < cnt:
#                 heappop(minh)
#                 heappush(minh, (cnt, i))
#         # 3. return the heap with val only
#         return [i for cnt, i in minh]